### PR TITLE
Cfy 2010 autoheal workflow

### PR DESCRIPTION
**Pull request by simon-bar**

### CODE REVIEW

File: source/guide/3.2/guide-authoring-policies.md [hunk 1]
```diff
@@ -40,6 +40,17 @@ Policy types are defined under the `policy_types` section of the blueprint.
 * The `source` attribute of a policy type can be either a relative path to a policy implementation within the blueprint directory or a URL to a policy hosted somewhere.
 * The `properties` attribute of a policy type defines the policy's properties schema. These properties are configured when instantiating policies within the `groups` section. In the following section we will describe how these properties can be used by a policy implementation.
 
+    The following properties are built-in and common for all policies:
+
+    * `policy_operates_on_group`:
+        Should the policy maintain it's state for for the whole group of nodes
```
Comment (dankilman) [inline source/guide/3.2/guide-authoring-policies.md]: it's --> its

File: source/guide/3.2/guide-authoring-policies.md [hunk 2]
```diff
@@ -40,6 +40,17 @@ Policy types are defined under the `policy_types` section of the blueprint.
...
+        or each node instance individually (false by default).
+    * `is_node_started_before_workflow`:
+        Before triggering workflow, check if the node state is started (true by default). This is a **very important check** - it prevents the `heal` workflow from getting executed after the built-in `uninstall` workflow takes place.
```
Comment (dankilman) [inline source/guide/3.2/guide-authoring-policies.md]: Before triggering workflow -> Before triggering a workflow

File: source/guide/3.2/guide-authoring-policies.md [hunk 3]
```diff
@@ -40,6 +40,17 @@ Policy types are defined under the `policy_types` section of the blueprint.
...
+    * `interval_between_workflows`:
+        Trigger workflow only if the last workflow was triggered earlier than interval-between-workflows seconds ago.
+        if the specified value is less than 0, workflows can run concurrently. The default value is 300 seconds.
```
Comment (dankilman) [inline source/guide/3.2/guide-authoring-policies.md]: if the specified -> If the specified

File: source/guide/3.2/guide-policies.md [hunk 4]
```diff
@@ -94,11 +94,88 @@ Built-in policies are declared in [`types.yaml`]({{page.types_yaml_link}}), whic
...
+                default: 0.5
+        source: https://raw.githubusercontent.com/cloudify-cosmo/cloudify-manager/master/resources/rest-service/cloudify/policies/ewma_stabilized.clj
 
 {% endhighlight %}
 
+* Host failure policy
+
+    This policy is based on intercepting expired events. The first monitoring event sent by a certain node instance will add this event type (determined by the event's service and host properties) to the policy engine index (Riemann's index mechanism). Later, if this type of event does not get sent for a period of 60 seconds for this particular node instance, it gets expired.
+
+    When an event expires and has been sent by a service contained in the list of services (policy's property) specified in the blueprint, the host failure policy passes this event to the restraints check and eventually processes triggers. It adds the `diagnose` field with value "heart-beat-failure" and the `failing_node` field with an id of the failing node to the event. Additionally, the event's state gets changed to "triggering_state".
```
Comment (dankilman) [inline source/guide/3.2/guide-policies.md]: eventually processes triggers -> eventually processes the policy triggers

File: source/guide/3.2/guide-policies.md [hunk 5]
```diff
@@ -94,11 +94,88 @@ Built-in policies are declared in [`types.yaml`]({{page.types_yaml_link}}), whic
...
+
+    This policy's implementation is based on [expired? Riemann function](http://riemann.io/api/riemann.streams.html#var-expired.3F).
+
+    You can find it on [github](https://github.com/cloudify-cosmo/cloudify-manager/blob/master/resources/rest-service/cloudify/policies/host_failure.clj).
+
+* Threshold policy
+
+    When for `stability_time` seconds all not expired events have metric that breach the `threshold`, the policy passes the last event to the restraints check and eventually processes triggers.
```
Comment (dankilman) [inline source/guide/3.2/guide-policies.md]: When for stability_time seconds all not expired events have metric that breach the threshold, the policy passes the last event to the restraints check and eventually processes triggers. -->
When for `stability_time` seconds, all none expired events have a `metric` value that breaches the threshold, the policy passes the last event to the restraints check and eventually processes the policy triggers.

File: source/guide/3.2/guide-policies.md [hunk 6]
```diff
@@ -21,7 +21,7 @@ diamond_package_ref: https://github.com/BrightcoveOS/Diamond
 Policies provide a way of analyzing a stream of events that correspond to a group of nodes (and their instances).
```
Comment (dankilman) [inline source/guide/3.2/guide-policies.md]: Let's update the example in this page to use the actual host_failure policy with its properties.
also, let's change the workflow names that is triggered from auto_heal to heal (in the example)

File: source/guide/3.2/guide-policies.md [hunk 7]
```diff
@@ -106,33 +183,24 @@ Built-in policies are not special in any way - they use the same API any other c
...
+Cloudify comes with a built-in support for auto healing different components.
 
-There are several things that need to be configured to work with auto healing. This guide will walk you through them, and explain the limitations the mechanism currently has.
+There are several things that need to be configured for auto healing to work. This guide will walk you through them and explain the mechanism's current limitations.
 
 In short, the process involves:
 
-* Configuring monitoring on compute nodes that should have auto healing.
+* Configuring monitoring on compute nodes that will be subject to autohealing.
 * Adding the `autoheal` workflow to the blueprint.
```
Comment (dankilman) [inline source/guide/3.2/guide-policies.md]: This section does not reflect that fact that heal is now a built-in workflow and should not be manually included by the user

File: source/guide/3.2/guide-policies.md [hunk 8]
```diff
@@ -106,33 +183,24 @@ Built-in policies are not special in any way - they use the same API any other c
...
-Make sure you don't miss the [Limitations](#limitations) section.
-{%endwarning%}
-
 ## Monitoring
 
 Add monitoring to compute nodes that require auto healing.
 
-At the moment, due to a limitation of existing policies, it is required that exactly one metric will be published by the monitored nodes.
-
-Another limitation requires that only compute nodes will have auto heal configured for them. Otherwise, when a certain compute node instance fails, node instances that are contained in it are also likely to fail, which in turn will cause the auto heal workflow to be triggered multiple times.
```
Comment (dankilman) [inline source/guide/3.2/guide-policies.md]: monitoring on host nodes only is still a limitation

File: source/guide/3.2/guide-policies.md [hunk 9]
```diff
@@ -168,12 +234,13 @@ workflows:
     mapping: default_workflows.cloudify.plugins.workflows.auto_heal_reinstall_node_subgraph
     parameters:
       node_id: {}
+      <!--TODO Default value -->
```
Comment (dankilman) [inline source/guide/3.2/guide-policies.md]: This entire workflow definition section can be removed

File: source/guide/3.2/guide-policies.md [hunk 10]
```diff
@@ -196,6 +263,12 @@ groups:
               # that was previously configured
               workflow: autoheal
```
Comment (dankilman) [inline source/guide/3.2/guide-policies.md]: rename autoheal to heal and remove the line stating it was previously configured

File: source/guide/3.2/guide-policies.md [hunk 11]
```diff
@@ -196,6 +263,12 @@ groups:
...
 
+              # Name of the service we want to short list (using regular expressions) and
+              # watch - every Diamond event has the service field set to some value.
+              # In our case, the ExampleCollector sends events with this value set to "example".
+              service:
```
Comment (dankilman) [inline source/guide/3.2/guide-policies.md]: The location of service in this example is wrong. It is not a trigger parameter but rather a policy property

File: source/guide/3.2/guide-policies.md [hunk 12]
```diff
@@ -211,22 +284,14 @@ groups:
                 diagnose_value: { get_property: [SELF, diagnose] }
 {% endhighlight %}
 
-In this example, we configured a group consisting of the `some_vm` node that was previously configured.
+In this example, we have configured a group consisting of the `some_vm` node specified before.
 
-We then configured a single `host_failure` policy for this group. The `host_failure` policy works by checking for expired events. The first monitoring event sent by a certain node instance will add this event type to the policy engine index (Riemann's index mechanism). After which, if this type of event was not sent for a period of 60 seconds for a certain node instance, it's considered a failed node instance and the policy triggers are executed.
+Afterwards, we configured a single `host_failure` policy for this group.
 
 The policy in the example has one `execute_workflow` policy trigger configured, which is mapped to execute the `autoheal` workflow.
```
Comment (dankilman) [inline source/guide/3.2/guide-policies.md]: execute the `heal` workflow

File: source/guide/3.2/guide-policies.md [hunk 13]
```diff
@@ -211,22 +284,14 @@ groups:
...
-
 {%warning title=Limitations%}
-* Only compute nodes can have auto healing configured for them.
-* Monitoring on these compute nodes can only generate a single metric type.
-* Each compute node must be specified in its own group in the `groups.members` section
-* Monitoring is set up when the deployment is created. As such, after the `uninstall` workflow has been called on a deployment, the `autoheal` workflow will probably be triggered after a while. At the moment, this is resolved by having the auto heal workflow check the state of the faling node host. If its state is different than `started` the workflow will end its execution without actually doing anything. This means that when the deployment executions are listed, you may find unexpected `autoheal` workflow executions in `terminated` state.
+* At the moment, this is resolved by having the auto heal workflow check the state of the faling node host. If its state is different than `started` the workflow will end its execution without actually doing anything. <!-- TODO -->
```
Comment (dankilman) [inline source/guide/3.2/guide-policies.md]: Only compute nodes can have auto healing configured for them.

File: source/guide/3.2/reference-builtin-workflows.md [hunk 14]
```diff
@@ -111,11 +111,11 @@ For each of the remaining node instances:
 
 **Workflow name:** *heal*
 
-**Workflow description:** The workflow for re-installing specific blueprint nodes.
+**Workflow description:** Reinstalls the whole subgraph of the system topology applying the `uninstall` and `install` workflows' logic respectively. The subgraph consists of all the nodes that are hosted in a compute containing the failing node and/or the compute itself. Additionally, this workflow handles unlinking and establishing all affected relationships in an appropriate order.
 
 **Workflow parameters:**
 
-  - *node_id*: The id of the failed node instance.
+  - *node_id*: The ID of the failing node that needs healing. The whole compute containing (or being) this node wil    l be reinstalled.
```
Comment (dankilman) [inline source/guide/3.2/reference-builtin-workflows.md]: why are there spaces here?

File: source/guide/3.2/guide-policies.md [hunk 15]
```diff
@@ -94,11 +94,88 @@ Built-in policies are declared in [`types.yaml`]({{page.types_yaml_link}}), whic
...
+            <<: *THRESHOLD_BASED_POLICY_PROPERTIES
+            ewma_timeless_r:
+                description: |
+                    r is the ratio between successive events. The smaller it is, the smaller impact on the computed value the most recent event has.
+                default: 0.5
+        source: https://raw.githubusercontent.com/cloudify-cosmo/cloudify-manager/master/resources/rest-service/cloudify/policies/ewma_stabilized.clj
 
 {% endhighlight %}
 
+* Host failure policy
```
Comment (dankilman) [inline source/guide/3.2/guide-policies.md]: Can we make the built-in policies documentation have a structure similar in nature to the structure in the built-in workflows section?
Also, can we provide a short yaml snippet that shows how each policy can be configured? Something similar in nature to what we do in the DSL specification section. (for example http://getcloudify.org/guide/3.2/dsl-spec-policy-types.html)(notice how this example contains both a "formal yaml" structure and also a specific example)

File: source/guide/3.2/guide-policies.md (hunk 8 above)
Comment (simon-bar) [inline source/guide/3.2/guide-policies.md]: I think it's only partly true - we can have a policy that, for instace, monitors if some process is working (using ps) and if it's not, we can heal the whole VM containing it. We only need to have an appropriate collector for this purpose. What we are completely unable to do is heal the node exclusively..

File: source/guide/3.2/reference-builtin-workflows.md (hunk 14 above)
Comment (simon-bar) [inline source/guide/3.2/reference-builtin-workflows.md]: VIM in action I guess ;)

File: source/guide/3.2/dsl-spec-policy-types.md [hunk 16]
```diff
@@ -44,21 +44,61 @@ Example:
 {%highlight yaml%}
 policy_types:
 
+  my_host_failure_policy:
```
Comment (dankilman) [inline source/guide/3.2/dsl-spec-policy-types.md]: The example is incorrect. properties in policy types are in schema structure, i.e:
``` yaml
properties:
       policy_operates_on_group: 
           description: The policy will maintain its state for each node instance individually.
           default: false
```
